src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,205 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> head;
+    private Node<T> tail;
+
+    static class Node<T> {
+        private T value;
+        private Node<T> next;
+        private Node<T> prev;
+
+        public Node(Node<T> prev, T value, Node<T> next) {
+            this.value = value;
+            this.prev = prev;
+            this.next = next;
+        }
+    }
+
     @Override
     public void add(T value) {
+        Node<T> newNode = new Node<>(tail, value, null);
+
+        if (head == null) {
+            head = newNode;
+        } else {
+            tail.next = newNode;
+        }
+        tail = newNode;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index < 0 || index > size()) {
+            throw new IndexOutOfBoundsException("Indeks jest poza zakresem listy: " + index);
+        }
+
+        if (index == 0) {
+            Node<T> newNode = new Node<>(null, value, head);
+            if (head != null) {
+                head.prev = newNode;
+            }
+            head = newNode;
+            if (tail == null) {
+                tail = newNode;
+            }
+        } else {
+            Node<T> current = head;
+            int currentIndex = 0;
+            while (current.next != null && currentIndex < index - 1) {
+                current = current.next;
+                currentIndex++;
+            }
+            if (current.next == null) {
+                Node<T> newNode = new Node<>(current, value, null);
+                current.next = newNode;
+                tail = newNode;
+            } else {
+                Node<T> newNode = new Node<>(current, value, current.next);
+                current.next.prev = newNode;
+                current.next = newNode;
+            }
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T value : list) {
+            add(value);
+        }
     }
 
     @Override
     public T get(int index) {
+        if (index < 0 || index >= size()) {
+            throw new IndexOutOfBoundsException("Indeks jest poza zakresem listy: " + index);
+        }
+
+        Node<T> current = head;
+        int currentIndex = 0;
+
+        while (current != null && currentIndex < index) {
+            current = current.next;
+            currentIndex++;
+        }
+        if (current != null && currentIndex == index) {
+            return current.value;
+        }
         return null;
     }
 
     @Override
     public T set(T value, int index) {
+        if (index < 0 || index >= size()) {
+            throw new IndexOutOfBoundsException("Indeks jest poza zakresem listy: " + index);
+        }
+
```

Comment:
Code duplication
All guard clauses are the same for each method besides `add`, extract them to a private utility checker method